server/src/main/java/com/hedera/block/server/persistence/storage/compression/AbstractCompression.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.block.server.persistence.storage.compression;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/**
+ * An abstract implementation of {@link Compression} that provides common
+ * functionality for all compression implementations. Used also to enforce
+ * compression API contract preconditions.
+ */
+abstract class AbstractCompression implements Compression {
+    /**
+     * Factory method, used to create the result for the
+     * {@link Compression#newCompressingOutputStream(Path)} method.
+     */
+    @NonNull
+    protected abstract OutputStream createNewCompressingOutputStream(@NonNull final Path pathToFile) throws IOException;
+
+    @NonNull
+    @Override
+    public final OutputStream newCompressingOutputStream(@NonNull final Path pathToFile) throws IOException {
+        final String errorMessage;
+        if (Files.isDirectory(pathToFile)) {
+            errorMessage = "The input path [%s] must not be a directory!".formatted(pathToFile);
+        } else if (Files.notExists(pathToFile.getParent())) {
+            errorMessage = "The path to the parent directory of the input path [%s] must exist!".formatted(pathToFile);
+        } else {
+            return createNewCompressingOutputStream(pathToFile);
+        }
+        throw new IllegalArgumentException(errorMessage);
+    }
```

Comment:
My whole idea for this abstract class is to ensure that preconditions that are part of the highest level api are always enforced. I want to make sure that it is not error prone to enforce preconditions, however I have a bit of a strange `code smell` feeling here. Opinions?

Reply:
I would expect `CompressingOutputStream` to wrap an `OutputStream`, not a `Path`.  This is assuming that all compressed streams are files; which is not always true.
At minimum we already know we must support S3 style bucket outputs; and we might need or want compression for quite a few other non-file options in the future.

Reply:
Hmm... yes, in addition to the comment above, I think this was the missing link on my side, assuming that the contract `Compression` is appropriate to generate streams based on a `Path`. In fact, now that I think of it, instead of the method to be called `newCompressingOutputStream`, I think it would be a better fit to name the method as `wrap` or similar, and as input, to accept an `OutputStream`. In that way I also decouple myself from the responsibility and the questionable contract behavior to _partially_ create the path to the destination (that is if the specific implementation will actually target a file on the filesystem...). Thanks for that! :)

Reply:
✅ Done. We have moved to another implementation.